### ISSUE TITLE
Traffic jam. street too short, doesn't match real intersection, vehicle turn conflicts

### ISSUE THREAD

**Issue by GoldenPhysics** [dabreegster/abstreet#129]
Map: Downtown Seattle
Problem location: Intersection of East Union and East Madison
![NoTurnBug](https://user-images.githubusercontent.com/18746766/86544139-ed9d1500-bf13-11ea-9ba9-dca13092c039.png)

The two cars in the middle on the very short street are each trying to turn left. Neither has moved for about 25 in-game minutes. This is causing a sizable traffic jam. 

Additionally, this generated intersection does not match what the real intersection looks like ([Google maps satellite view](https://www.google.com/maps/@47.6129086,-122.3166021,90m/data=!3m1!1e3)) but I don't know how it's possible for map generation to know this based on the [OpenStreetMap data](https://www.openstreetmap.org/#map=19/47.61294/-122.31678). I'll just note that mismatch here and not go in to more detail. If more detail is desired, I think another issue should be made for that.

I can think of a few possible solutions off the top of my head of unknown quality.
- Set a minimum distance a street segment must have, which would lengthen the segment that is too short when the map is generated. Potential problems: What do you move to lengthen the street? Do you lengthen it enough for a bus? Will that make other geometry overlap?
- If intersections are too close together, combine them when the map is generated. Potential problems: Street alignment? Matching reality? In this case, combining them would match reality, but that may not be true for all cases, particularly places with split one way streets.
- If a car has waited for some amount of time (10 in-game minutes? and are at the front of an intersection queue? 10 minutes at the front of the queue?) allow them to move despite geometry conflicts (probably a good idea to log this, too). Potential problems: How much time? Does that just hide the problem from the player?

**Comment by dabreegster** [comment]
Thanks for reporting! This is one of those hard problems I've been struggling with since the beginning. I now consider it two separate problems:

First, the geometry is bad, which makes it look strange, makes it really tough to edit anything in the area, and which can mess up turn conflicts (two movements conflict if they cross, but when the geometry is this messed up, the movements are kind of bogus). Your first two ideas address this problem. I think the first wouldn't work; in the example here, there's not even room to curve E Union St and lengthen that tiny north/south segment of 12th without overlapping buildings.

I've attempted variations of the second idea before with various levels of success, including manually deleting the OpenStreetMap data locally and trying to draw it again. In this particular example, I can imagine some changes to the [algorithm](https://docs.google.com/presentation/d/1cF7qFtjAzkXL_r62CjxBvgQnLvuQ9I2WTE2iX_5tMCY/edit#slide=id.g6201f22714_0_16) that finds the intersection polygon. When there's a short road after the first pass of the algorithm, see if the two intersections on either end of it physically overlap any of the roads connected to the opposite intersection. If so, repeat the algorithm, treating the first two attempts at polygons as extra things to generate collisions, and trim back further. Some of the difficulty with this isn't just producing a better polygon, but dealing with all of the turns between connected roads. When you delete a road in the middle, there are some weird edge cases that can happen.

The second problem is how vehicles get stuck when the geometry is this weird. I did some work on this recently that I haven't written up anywhere yet; in short, I'll experiment with adding this intersection to [these overrides](https://github.com/dabreegster/abstreet/blob/7d4cf4cd5b27bce37cf35d3d8aecf0e3d5148a2f/sim/src/mechanics/intersection.rs#L714) to see if we can temporarily make some progress. The more robust fix is to treat the two intersections as one at the simulation layer, meaning that agents would acquire a lock on a turn->lane->turn movement as one, instead of just a turn. Then nobody could ever get stuck in a short lane.

I'll start thinking about this again and trying stuff soon. Ideas welcome!
